### PR TITLE
Third milestone

### PR DESCRIPTION
**Pull request by rayking12**
Things to look at for:
- User can now send friendship invitation
- Mutual friendships
-  friendship table

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Invalid PR 😞 
Nice work setting up the project for this milestone. Unfortunately, this PR is invalid for the following reasons:
- [x] There's no description for this PR. Add a detailed description of what features are covered in this PR. For example, a user can send a friend request to another user, etc. The description helps to provide a high level overview of what this PR is about.
- [x] Enable `CI/CD` with `Github actions`. This is a hard requirement for any PR.
- [x] Submit for another review after making the changes.


### Martin Afani

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## Hi @rayking12 and @Oxford-G  :wave: 

You've done well so far :+1:  and I commend you for that :clap: 
However, there are still some changes to be made before this milestone can be approved.

## REQUIRED

* [x] UsersController#index displays all users with `Friend Request` button next to users who are not their friends already ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone3/userctr.png) from [Odin Project](https://www.theodinproject.com/courses/ruby-on-rails/lessons/final-project))
* [x] PostsController#index displays posts in a timeline (their own posts + posts from their friends) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/timeline.png) from [Odin Project](https://www.theodinproject.com/courses/ruby-on-rails/lessons/final-project))

### Make the required changes and request another review.
## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:

### To Highlight: :angel: 

- [ ]  Good Pull request description. :+1: 
- [ ] Great job implementing the requested changes. :muscle: 
- [ ] It is perfect that you are using helper in order to remove all logic from views. :1st_place_medal: 

### To Improve :factory_worker:

- [x] Fix linter errors:


- [x] Some of your Unit and Integration tests failed:


- [x] Friendship model:  when friendship is confirmed, an additional row with reversed user-friend pair is created
```
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] User model: thanks to this, we can find user friends only by user_id in friendships table
```
# User model
has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [x] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory:
```
# User model
# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

- [x] User#friend_requests can be defined as association with SQL query instead of processing objects in application memory:
```
# User model
# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [x] Extend your unit tests.


That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello King Ray & Gerald Enekwechi** 
Good job with the improvements for this project :muscle:  :muscle:
Excelent work implementing counter for posts' comments and likes :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [x] Fix rubocop offenses
- [x] When a friendship is requested a row is created in the friendship table with status confirmed equals false. When that friendship is confirmed then a second row is created, both rows then have a confirmed state that equals true. That's the expected behavior for this milestone.
- [x] Make sure your project passes all your tests:



- [x] In your tests, use only 1 "expect" for each "it"
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Comment by Oxford-G** [comment]
The unit and integration tests are working perfectly in our local machine. Also we have corrected the Rubocop errors

File: app/models/friendship.rb
Comment (enaburekhan) [inline app/models/friendship.rb]: line 7 is not clear. I can't locate where you initially defined it. Did you mean :user.id_friend.id?

File: app/models/friendship.rb
Comment (enaburekhan) [inline app/models/friendship.rb]: - [x] **when friendship is confirmed, an additional row with reversed user-friend pair is created**. Just like the lasts two reviewers already highlighted. we need to write a method to confirm friend inside the friendship class. Check the method that was initially given and produce something similar,

File: app/models/user.rb
Comment (enaburekhan) [inline app/models/user.rb]: Writing it this way using ruby is good. But, in this milestone, we want you to use SQL query.  This will allow us to find user friends only by user_id in friendships table. Your code should be something like this:

has_many :confirmed_something, -> { where confirmed: true }, class_name: "Something"
has_many :things, through: :confirmed_something.

File: app/models/user.rb
Comment (enaburekhan) [inline app/models/user.rb]: - [x] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby). Rewrite it to use SQL query. Now you understand better why we prefer SQL query:blush:.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes requested:full_moon:
**Hi team!
Great job on this milestone:clap::clap:.** 

**Major highlights**

- [x] It has a simple method to get all posts written by user on their friends

- [x] It uses SQL query instead of ruby method to create friend requests

However, we still need to make a few changes to get this project approved:muscle:.

**To improve**

- [x] Kindly go through the PR below to see my comments. Make the required changes and corrections. 

- [x] Currently, we only wrote tests for the model. We also need to write for the controller as well. Both unit and integration tests that covers both positive and negative test case scenarios.

Make the required changes and when you are ready, request another review.
Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @rayking12 and @Oxford-G, Great job! 👏  Both of you have done well on this project. Nice job on the specs.

However, there are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.

- [x] Your commit messages should be capitalized, and written in the imperative tense. You can watch this [video](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6) on how to edit your commit messages using git rebase
 See [Git Commit Best Practices](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) for more insights

- [x] While making some changes you may have broken your code, clicking on See Profile produces an error
Screen Shot 2021-03-08 at 3 56 03 PM

- [x] I should not be able to send myself a friend request.

- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
See the affected file below:
> `layouts/application.html.erb`
**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by acushlakoncept — APPROVED** [review]
# Status: PROJECT APPROVED  🎉 🚀

### Hi, @rayking12 and @Oxford-G  congratulations on completing this milestone and taking the time to implement the requested changes.

I will go ahead and mark this as complete, feel free to merge when you are ready. All the best with the next milestone.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Pull request merged**
